core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java

```diff
@@ -413,6 +442,9 @@ public String source() {
         }
 
         assert batchResult.executionResults != null;
+        assert batchResult.executionResults.size() == toExecute.size();
+        final ClusterStateTaskExecutor.BatchResult<T> finalBatchResult = batchResult;
+        assert toExecute.stream().map(updateTask -> updateTask.task).allMatch(finalBatchResult.executionResults::containsKey);
```

Comment:
that will be hard to debug if it's false - which one misses? (holds for the one above as well).